aten/src/ATen/native/TensorAdvancedIndexing.cpp

```diff
@@ -68,7 +69,99 @@
 #include <numeric>
 #include <vector>
 
-namespace at { namespace native {
+namespace at {
+namespace meta {
+
+native::SCATTER_GATHER_OP get_operator_enum(const c10::string_view reduce) {
+  if (reduce == "add") {
+    return native::SCATTER_GATHER_OP::REDUCE_ADD;
+  } else if (reduce == "multiply") {
+    return native::SCATTER_GATHER_OP::REDUCE_MULTIPLY;
+  } else {
+    TORCH_CHECK(false, "reduce argument must be either add or multiply.");
+  }
+}
+
+template <typename Meta>
+void scatter_meta_impl(
+    Meta& meta,
+    const Tensor& self,
+    int64_t dim,
+    const Tensor& index,
+    const c10::optional<Tensor>& src = nullopt,
+    const c10::optional<c10::string_view> reduce = nullopt) {
+  int64_t wrapped_dim = at::maybe_wrap_dim(dim, self.dim());
+  at::native::scatter_gather_dtype_check("scatter", self, index, src);
+  at::native::scatter_shape_check(self, wrapped_dim, index, src);
+
+  meta.set_output(self.sizes(), self.options());
+
+  auto output = meta.maybe_get_output(0);
+  at::assert_no_internal_overlap(output);
```

Comment:
Yeah OK, this is acceptable. It is a little wasteful when the output is freshly allocated, so a very modest improvement you could do is swing this line before the `set_output` call and test if `output` is defined, and only do the overlap tests if it is defined (if it is not defined, it means we are going to freshly allocate and you don't need to test for overlap as it is guaranteed not to occur), but I'm not going to block the PR on this if there are no other substantive problems.

Reply:
You are right. I didn't really stop to think about the complexity of these checks.
Since I have no strong reasons not to do so, I will implement that change!